src/node/txreconciliation.cpp

```diff
@@ -0,0 +1,192 @@
+{
+    // According to BIP-330, salts should be combined in ascending order.
+    uint64_t salt1 = local_salt, salt2 = remote_salt;
+    if (salt1 > salt2) std::swap(salt1, salt2);
+
+    static const auto RECON_SALT_HASHER = TaggedHash(RECON_STATIC_SALT);
+    return (CHashWriter(RECON_SALT_HASHER) << salt1 << salt2).GetSHA256();
+}
+
+/**
+ * Used to keep track of the ongoing reconciliations, the transactions we want to announce to the
+ * peer when next transaction reconciliation happens, and also all parameters required to perform
+ * reconciliations.
+ */
+class ReconciliationState {
+
+    public:
+
+    /**
+     * Reconciliation involves exchanging sketches, which efficiently represent transactions each
+     * peer wants to announce. Sketches are computed over transaction short IDs.
+     * These values are used to salt short IDs.
+     * TODO: These are currently public to overcome compiler warning (-Wunused-private-field).
+     *       They should be made private once they are used in one of the following commits.
+     */
+    const uint64_t m_k0, m_k1;
+
+    /**
+     * Reconciliation protocol assumes using one role consistently: either a reconciliation
+     * initiator (requesting sketches), or responder (sending sketches). This defines our role.
+     * */
+    const bool m_we_initiate;
+
+    ReconciliationState(uint64_t k0, uint64_t k1, bool we_initiate) :
+        m_k0(k0), m_k1(k1), m_we_initiate(we_initiate) {}
+};
+
+} // namespace
+
+/** Actual implementation for TxReconciliationTracker's data structure. */
+class TxReconciliationTracker::Impl {
+
+    mutable Mutex m_mutex;
+
+    /**
+     * Per-peer salt is used to compute transaction short IDs, which will be later used to
+     * construct reconciliation sketches.
+     * Salt is generated randomly per-peer to prevent:
+     * - linking of network nodes belonging to the same physical node
+     * - halting of relay of particular transactions due to short ID collisions (DoS)
+     */
+    std::unordered_map<NodeId, uint64_t> m_local_salts GUARDED_BY(m_mutex);
+
+    /**
+     * Keeps track of ongoing reconciliations with a given peer.
+     */
+    std::unordered_map<NodeId, ReconciliationState> m_states GUARDED_BY(m_mutex);
+
+    public:
+
+    std::tuple<bool, bool, uint32_t, uint64_t> PreRegisterPeer(NodeId peer_id, bool peer_inbound)
+    {
+        bool we_initiate_recon, we_respond_recon;
+        // Currently reconciliation roles are defined by the connection direction: only the inbound
+        // peer initiate reconciliations and the outbound peer is supposed to only respond.
+        // If the way we assign reconciliation roles is updated, RECON_VERSION should be bumped.
+        if (peer_inbound) {
+            we_initiate_recon = false;
+            we_respond_recon = true;
+        } else {
+            we_initiate_recon = true;
+            we_respond_recon = false;
+        }
+
+        uint64_t m_local_recon_salt(GetRand(UINT64_MAX));
```

Comment:
The reason for computing it here is that we need to return the salt from `PreRegisterPeer`, because we send the salt to the peer after pre-registering it.

Reply:
Right, didn't think of that, thanks for clearing up!